README.md
# gols

**Installation**

* `git clone https://github.com/euri10/gols.git`
* either use a virtual env or your distro packages, I use python3 and didn't test on python2 yet, but  want to learn Tox more so this is a work in progress
* if you use a venv please refer to my blog post below on the pygobject dependency, or see [this](http://stackoverflow.com/a/31609727/3581357)
* run `python3 setup.py install`, or if you prefer `pip install -e .` within your venv
* you can now run gols as a cli, `gols --help` will show you the commands and subsequent options, your watch needs to be mounted
* should you want to automate the upload a little more by triggering the script automatically, please refer to my blog post [here](https://medium.com/@euri10/gols-garmin-on-linux-sucks-f1f065f7529a#.3htyjn6q8)
or install the following systemd unit after you created the correct /etc/fstab entry

```bash
#garmin fenix 2
UUID=489A-9E97 /media/fenix2 vfat auto,nofail,rw,user,uid=1000,gid=1000 0 2
```

You'll get the UUID running `sudo blkid`.
The important parts of the entry are `auto` to tell systemd to mount the filesystem as soon as the device is available and `nofail` to ensure that boot does not fail when the device is not available (you've been warned)
(see [here](https://ddumont.wordpress.com/2016/04/24/automount-usb-devices-with-systemd/))
`rw,user,uid=1000,gid=1000` express the fact the mount will be under read-write and that the user who will run the script will have the correct permissions

```systemd
[Unit]
Description=gols a little less now
# systemctl --user list-units '*mount' might give you the exact unit name for 
# lines below
Requires=media-fenix2.mount
After=media-fenix2.mount
[Service]
# the path of a bash script that will launch the python cli with the command
# options you want, see example below
ExecStart=/home/user/PycharmProjects/gols/gols.sh
[Install]
WantedBy=media-fenix2.mount
```
Finally a bash script where you specify what upload options you want to trigger

```bash
#!/bin/bash
/home/user/venv/gols/bin/python /home/user/PycharmProjects/gols/gols/gols.py upload -d /media/fenix2/Garmin/Activity -n -m
```

```bash
(gols) ➜  gols git:(master) ✗ gols
Usage: gols [OPTIONS] COMMAND [ARGS]...

Options:
  --debug / --no_debug  Set to true to see debug logs on top of info
  --help                Show this message and exit.

Commands:
  upload  uploads .fit files to your garmin connect account
  ```
  
**Issues**

More than happy to try to help, this isn't extensively tested, other distro may have other settings I don't know, I'm on Debian testing by the way

CHANGELOG.rst

Changelog
=========

1.1.2 (unreleased)
------------------
* Change login process to new API


0.1.0 (2017-09-06)
------------------

* First release on PyPI.


tests/test_gols.py

from click.testing import CliRunner

from gols.cli import main


def test_main():
    runner = CliRunner()
    result = runner.invoke(main, [])

    assert result.output == '()\n'
    assert result.exit_code == 0


tests/test_config.py
import os.path

import pytest
from click.testing import CliRunner

from src.gols import Config, cli


@pytest.fixture()
def t_config():
    t = Config()

    return t


def test_config_created(t_config):
    runner = CliRunner()
    result = runner.invoke(cli(t_config), )


tests/test_upload.py
import os

import pytest
from click.testing import CliRunner
from testfixtures import LogCapture

from gols.cli import main
from gols.cli import upload


@pytest.fixture(scope='function')
def runner():
    yield CliRunner()


@pytest.fixture(scope='function')
def fs():
    if not os.path.exists(os.path.join(os.path.dirname(os.path.dirname(__file__)), 'tests', 'directory_fit')):
        os.makedirs(os.path.join(os.path.dirname(os.path.dirname(__file__)), 'tests', 'directory_fit'))
    yield os.path.join(os.path.dirname(os.path.dirname(__file__)), 'tests', 'directory_fit')


@pytest.fixture(scope='function')
def cdf():
    if not os.path.exists(os.path.join(os.path.dirname(os.path.dirname(__file__)), 'tests', 'conf_dir_fit')):
        os.makedirs(os.path.join(os.path.dirname(os.path.dirname(__file__)), 'tests', 'conf_dir_fit'))
    yield os.path.join(os.path.dirname(os.path.dirname(__file__)), 'tests', 'conf_dir_fit')


def test_debug_turned_on(runner):
    with LogCapture() as lc:
        result = runner.invoke(main, ['--debug', 'upload'])
        assert result.exit_code == 2
        assert 'Debug level set on' in str(lc)


def test_debug_turned_off(runner):
    with LogCapture() as lc:
        result = runner.invoke(main, ['upload'])
        assert result.exit_code == 2
        assert 'Info level set on' in str(lc)


def test_upload_help(runner):
    result = runner.invoke(upload, ['--help'])
    assert result.exit_code == 0
    assert 'Usage' in result.output


def test_required_fit_directory(runner, fs):
    result = runner.invoke(upload, ['upload', fs])
    assert result.exit_code == 2
    assert 'Error: Missing option' in result.output


def test_upload_fit(runner, fs, cdf):
    username = 'gols@mailinator.com'
    password = 'G0lsG0ls'
    with LogCapture() as lc:
        # logger = logging.getLogger()
        result = runner.invoke(main, ['--debug', 'upload', '-d', fs, '-c', cdf, '-u', username, '-p', password])  # noqa
        print(result.output)
        print(lc)
        assert result.exit_code == 0
        assert 'Done uploading' in str(lc)


src/gols.py
import os
import logging
import click
import requests
import shutil
import yaml


logger = logging.getLogger(__name__)
logging.basicConfig()


class Config(dict):
    def __init__(self):
        self.template = {'username': 'username', 'password': 'password'}
        self.path = self._create_or_load(click.get_app_dir('gols'),
                                         'config.yaml')
        self.conf_dir_fit = self._conf_dir_fit(click.get_app_dir('gols'), 'fit')
        super(Config, self).__init__()

    @staticmethod
    def _conf_dir_fit(app_directory, directory):
        conf_dir_fit = os.path.join(app_directory, directory)
        if not os.path.exists(conf_dir_fit):
            os.makedirs(conf_dir_fit)
        return conf_dir_fit

    @staticmethod
    def _create_or_load(directory, fn):
        if os.path.exists(os.path.join(directory, fn)):
            return os.path.join(directory, fn)
        else:
            if not os.path.exists(directory):
                os.makedirs(directory)
                with open(os.path.join(directory, fn), 'w') as ymlfile:
                    yaml.dump({'username': 'username', 'password': 'password'},
                              ymlfile, default_flow_style=False)
                return os.path.join(directory, fn)

    def load(self):
        with open(self.path, 'r') as ymlfile:
            self.configdict = yaml.load(ymlfile)
            if self.configdict == self.template:
                if click.confirm('the config file {} has been created, edit it please'.format(self.path), abort=True):
                    with open(self.path, 'w') as ymlfile:
                        username = click.prompt(
                            'enter you Garmin Connect username')
                        password = click.prompt(
                            'enter you Garmin Connect password',
                            hide_input=True)
                        yaml.dump({'username': username, 'password': password},
                                  ymlfile, default_flow_style=False)
                    self.load()


pass_config = click.make_pass_decorator(Config, ensure=True)


@click.group()
@pass_config
@click.option('--debug/--no_debug', default=False,
              help='Set to true to see debug logs on top of info')
def cli(config, debug):
    config.load()
    if debug:
        from http.client import HTTPConnection
        HTTPConnection.debuglevel = 1
        requests_log = logging.getLogger("requests.packages.urllib3")
        requests_log.setLevel(logging.DEBUG)
        requests_log.propagate = True
        logging.root.setLevel(level=logging.DEBUG)
    else:
        logger.setLevel(level=logging.INFO)


@click.command(short_help='uploads .fit files to your garmin connect account')
@pass_config
@click.option('--directory_fit', '-d', required=True,
              type=click.Path(exists=True, file_okay=False),
              help='Path of your .fit files on your watch mount path')
@click.option('--notification/--no_notification', '-n', default=False,
              help='Get notified')
@click.option('--move/--no_move', '-m', default=False,
              help='Move files upon upload')
def upload(config, directory_fit, notification, move):
    username = config.configdict['username']
    password = config.configdict['password']
    conf_dir_fit = config.conf_dir_fit

    headers = {
        'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64; rv:51.0) Gecko/20100101 Firefox/51.0',
    }
    params_login = {
        'service': 'https://connect.garmin.com/modern/',
        'webhost': 'olaxpw-conctmodern011',
        'source': 'https://connect.garmin.com/en-US/signin',
        'redirectAfterAccountLoginUrl': 'https://connect.garmin.com/modern/',
        'redirectAfterAccountCreationUrl': 'https://connect.garmin.com/modern/',
        'gauthHost': 'https://sso.garmin.com/sso',
        'locale': 'en_US',
        'id': 'gauth-widget',
        'cssUrl': 'https://static.garmincdn.com/com.garmin.connect/ui/css/gauth-custom-v1.2-min.css',
        'clientId': 'GarminConnect',
        'rememberMeShown': 'true',
        'rememberMeChecked': 'false',
        'createAccountShown': 'true',
        'openCreateAccount': 'false',
        'usernameShown': 'false',
        'displayNameShown': 'true',
        'consumeServiceTicket': 'false',
        'initialFocus': 'true',
        'embedWidget': 'false',
        'generateExtraServiceTicket': 'false',
        'globalOptInShown': 'false',
        'globalOptInChecked': 'false',
        'connectLegalTerms': 'true',
    }
    data_login = {
        'username': username,
        'password': password,
        'embed': 'true',
        'lt': 'e1s1',
        '_eventId': 'submit',
        'displayNameRequired': 'false',
        'rememberme': 'on',
    }
    # begin session with headers because, requests client isn't an option,
    # dunno if Icewasel is still banned...
    logger.info('Login into Garmin connect')
    s = requests.session()
    s.headers.update(headers)
    # we need the cookies from the login page before we can post the user/pass
    url_login = 'https://sso.garmin.com/sso/login'
    req_login = s.get(url_login, params=params_login)
    if req_login.status_code != 200:
        logger.info('issue with {}, you can turn on debug for more info'.format(
            req_login))
    req_login2 = s.post(url_login, data=data_login)
    if req_login2.status_code != 200:
        logger.info('issue with {}, you can turn on debug for more info'.format(
            req_login2))
    # we need that to authenticate further,
    # kind like a weird way to login but...
    t = req_login2.cookies.get('CASTGC')
    t = 'ST-0' + t[4:]
    # now the auth with the cookies we got
    # url_post_auth = 'https://connect.garmin.com/modern'
    # this one I still don't know how to get it
    url_post_auth = 'https://connect.garmin.com/post-auth/login'
    params_post_auth = {'ticket': t}
    req_post_auth = s.get(url_post_auth, params=params_post_auth)
    if req_post_auth.status_code != 200:
        logger.info('issue with {}, you can turn on debug for more info'.format(
            req_post_auth))
    logger.info('Let\'s upload stuff now')
    # login should be done we upload now

    url_upload = 'https://connect.garmin.com/modern/proxy/upload-service/upload/.fit'
    if len(os.listdir(directory_fit)):
        for filename in os.listdir(directory_fit):
            logger.info('uploading:  {}'.format(filename))
            files = {'data': (filename,
                              open(os.path.join(directory_fit, filename), 'rb'),
                              'application/octet-stream')
                     }
            s.headers.update({'Referer':'https://connect.garmin.com/modern/import-data', 'NK':'NT'})
            req5 = s.post(url_upload, files=files)
            if req5.status_code != 200:
                logger.info(
                    'issue with {}, you can turn on debug for more info'.format(
                        req5))

            fn = req5.json()['detailedImportResult']['fileName']
            if 'failures' in req5.json()['detailedImportResult']:
                for failure in req5.json()['detailedImportResult']['failures']:
                    m_failures = failure['messages'][0]['content']
                    logger.info(m_failures)
                    if notification:
                        pass
            if 'successes' in req5.json()['detailedImportResult']:
                for successes in req5.json()['detailedImportResult']['successes']:
                    m_success = 'https://connect.garmin.com/modern/activity/' + str(
                        successes['internalId'])
                    logger.info(m_success)
                    if notification:
                        pass
            if move:
                shutil.move(os.path.join(directory_fit, filename),
                            os.path.join(conf_dir_fit, filename))
        logger.info('Done uploading')
    else:
        logger.info('No file found in {}'.format(directory_fit))
    logger.info('Finished')


cli.add_command(upload)

if __name__ == '__main__':
    cli()


src/gols/cli.py
"""
Module that contains the command line app.

Why does this file exist, and why not put this in __main__?

  You might be tempted to import things from __main__ later, but that will cause
  problems: the code will get executed twice:

  - When you run `python -mgols` python will execute
    ``__main__.py`` as a script. That means there won't be any
    ``gols.__main__`` in ``sys.modules``.
  - When you import __main__ it will get executed again (as a module) because
    there's no ``gols.__main__`` in ``sys.modules``.

  Also see (1) from http://click.pocoo.org/5/setuptools/#setuptools-integration
"""

import logging
import os
import re
import shutil

import click
import requests
from bs4 import BeautifulSoup

logger = logging.getLogger(__name__)
logging.basicConfig()


@click.group()
@click.option('--debug/--no_debug', default=False,
              help='Set to true to see debug logs on top of info')
def main(debug):
    if debug:
        requests_log = logging.getLogger("requests.packages.urllib3")
        requests_log.setLevel(logging.DEBUG)
        requests_log.propagate = True
        logging.root.setLevel(level=logging.DEBUG)
        logger.info('Debug level set on')
    else:
        logger.setLevel(level=logging.INFO)
        logger.info('Info level set on')


@main.command(short_help='uploads .fit files to your garmin connect account')
@click.option('--directory_fit', '-d', required=True,
              type=click.Path(exists=True, file_okay=False),
              help='Path of your .fit files on your watch mount path')
@click.option('--move/--no_move', '-m', default=False,
              help='Move files upon upload')
@click.option('--username', '-u', required=True,
              default=lambda: os.environ.get('GARMINCONNECT_USERNAME', ''),
              help='The GARMINCONNECT_USERNAME environment variable should you have one set')  # noqa
@click.option('--password', '-p', required=True,
              default=lambda: os.environ.get('GARMINCONNECT_PASSWORD', ''),
              help='The GARMINCONNECT_PASSWORD environment variable should you have one set ')  # noqa
@click.option('--conf_dir_fit', '-c', required=True,
              type=click.Path(exists=True, file_okay=False, dir_okay=True, writable=True, readable=True))  # noqa
def upload(directory_fit, move, username, password, conf_dir_fit):
    logger.info('Uplading stuff')

    WEBHOST = "https://connect.garmin.com"
    REDIRECT = "https://connect.garmin.com/modern/"
    BASE_URL = "https://connect.garmin.com/signin/"
    # GAUTH = "http://connect.garmin.com/gauth/hostname"
    SSO = "https://sso.garmin.com/sso"
    CSS = "https://static.garmincdn.com/com.garmin.connect/ui/css/gauth-custom-v1.2-min.css"

    headers = {
        'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64; rv:66.0) Gecko/20100101 Firefox/66.0',
        'Origin': "https://sso.garmin.com"
    }

    params = {'service': REDIRECT,
              'webhost': WEBHOST,
              'source': BASE_URL,
              'redirectAfterAccountLoginUrl': REDIRECT,
              'redirectAfterAccountCreationUrl': REDIRECT,
              'gauthHost': SSO,
              'locale': 'en_US',
              'id': 'gauth-widget',
              'cssUrl': CSS,
              'clientId': 'GarminConnect',
              'rememberMeShown': 'true',
              'rememberMeChecked': 'false',
              'createAccountShown': 'true',
              'openCreateAccount': 'false',
              'displayNameShown': 'false',
              'consumeServiceTicket': 'false',
              'initialFocus': 'true',
              'embedWidget': 'false',
              'generateExtraServiceTicket': 'true',
              'generateTwoExtraServiceTickets': 'false',
              'generateNoServiceTicket': 'false',
              'globalOptInShown': 'true',
              'globalOptInChecked': 'false',
              'mobile': 'false',
              'connectLegalTerms': 'true',
              'locationPromptShown': 'true',
              'showPassword': 'true'}

    data_login = {
        'username': username,
        'password': password,
        'embed': 'false'
    }

    # begin session with headers because, requests client isn't an option, dunno if Icewasel is still banned...
    logger.info('Login into Garmin connect')
    s = requests.session()
    s.headers.update(headers)
    # we need the cookies and csrf token from the login page before we can post the user/pass
    url_gc_login = 'https://sso.garmin.com/sso/signin'
    req_login = s.get(url_gc_login, params=params, headers=headers)
    if req_login.status_code != 200:
        logger.info('issue with {}, you can turn on debug for more info'.format(
            req_login))

    csrf_input = BeautifulSoup(req_login.content.decode(), 'html.parser').find('input', {'name': '_csrf'})
    if not csrf_input or not csrf_input.get('value'):
        raise Exception('Unable to get csrf token from login page.')
    data_login['_csrf'] = csrf_input.get('value')

    req_login2 = s.post(url_gc_login, data=data_login, params=params, headers=headers)
    if req_login2.status_code != 200:
        logger.info('issue with {}, you can turn on debug for more info'.format(
            req_login2))
    # extract the ticket from the login response
    pattern = re.compile(r".*\?ticket=([-\w]+)\";.*", re.MULTILINE | re.DOTALL)
    match = pattern.match(req_login2.content.decode())
    if not match:
        raise Exception('Did not get a ticket in the login response. Cannot log in. Did you enter the correct username and password?')
    login_ticket = match.group(1)
    print('login ticket=' + login_ticket)

    url_gc_post_auth = 'https://connect.garmin.com/modern/activities?'

    params_post_auth = {'ticket': login_ticket}
    req_post_auth = s.get(url_gc_post_auth, params=params_post_auth)
    if req_post_auth.status_code != 200:
        logger.info('issue with {}, you can turn on debug for more info'.format(
            req_post_auth))
    logger.info('Let\'s upload stuff now')
    # login should be done we upload now

    # url_upload = 'https://connect.garmin.com/proxy/upload-service-1.1/json/upload/.fit'
    url_upload = 'https://connect.garmin.com/modern/proxy/upload-service/upload/.fit'
    if len(os.listdir(directory_fit)):
        logger.debug([f for f in os.listdir(directory_fit) if os.path.isfile(os.path.join(directory_fit, f))])
        for filename in [f for f in os.listdir(directory_fit) if os.path.isfile(os.path.join(directory_fit, f))]:
            logger.info('uploading:  {}'.format(filename))
            files = {'data': (filename,
                              open(os.path.join(directory_fit, filename), 'rb'),
                              'application/octet-stream')
                     }
            s.headers.update({'Referer': 'https://connect.garmin.com/modern/import-data', 'NK': 'NT'})
            req5 = s.post(url_upload, files=files)
            if req5.status_code != 201:
                logger.info(
                    'issue with {}, you can turn on debug for more info'.format(
                        req5))

            # fn = req5.json()['detailedImportResult']['fileName']
            if 'failures' in req5.json()['detailedImportResult']:
                for failure in req5.json()['detailedImportResult']['failures']:
                    m_failures = failure['messages'][0]['content']
                    logger.info(m_failures)
            if 'successes' in req5.json()['detailedImportResult']:
                for successes in req5.json()['detailedImportResult']['successes']:
                    m_success = 'https://connect.garmin.com/modern/activity/' + str(
                        successes['internalId'])
                    logger.info(m_success)

            if move:
                shutil.move(os.path.join(directory_fit, filename),
                            os.path.join(conf_dir_fit, filename))

        logger.info('Done uploading')
    else:
        logger.info('No file found in {}'.format(directory_fit))
    logger.info('Finished')
